queue_consumer.go

```diff
@@ -132,30 +151,48 @@ func (mf *Consumer) Run(ctx context.Context) error {
 	del := NewBatchDeleter(cleanupCtx, cleanupWG, mf.s, mf.DeleteMessageAccumulatorTimeout, mf.DeleteMessageDrainTimeout)
 
 	messages := make(chan job)
-	go mf.receiveMessages(cleanupCtx, cleanupWG, messages, del)
-	var done bool
+	go mf.receiveMessages(cleanupCtx, cleanupWG, mf.shutDown, messages, del)
+	defer func() {
+		close(jobs)
+		wg.Wait()
+		cleanupCancel()
+		cleanupWG.Wait()
+	}()
+
 	for {
 		// Stop if the context was cancelled
 		select {
 		case <-ctx.Done():
-			close(jobs)
-			wg.Wait()
-			cleanupCancel()
-			cleanupWG.Wait()
 			return ctx.Err()
-		case msg := <-messages:
-			if done {
-				break
```

Comment:
I bet this was a bug, since it breaks from the `select`, not the enclosing `for`.

Reply:
Ha, interesting.  It looks like it was a way to force the `<-messages` branch to be a no-op once the context was `Done`.